### PR TITLE
[Merged by Bors] - feat(Algebra/Category/Ring/Constructions): categorical product of rings is the cartesian product

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/Ring/Constructions.lean [hunk 1]
```diff
@@ -215,6 +215,53 @@ set_option linter.uppercaseLean3 false in
...
+  fac := by aesop
+  uniq s g h := FunLike.ext _ _ fun x ↦ funext fun i ↦ FunLike.congr_fun (h ⟨i⟩) x
+
+/--
+The categorical product and the usual product agrees
+-/
+def piIsoPi : ∏ R ≅ CommRingCat.of ((i : ι) → R i) :=
+  limit.isoLimitCone ⟨_, piFanIsLimit R⟩
+
+
```
Comment (erdOne) [inline Mathlib/Algebra/Category/Ring/Constructions.lean:253]:
```suggestion

```

File: Mathlib/Algebra/Category/Ring/Constructions.lean [hunk 2]
```diff
@@ -215,6 +215,53 @@ set_option linter.uppercaseLean3 false in
...
+/--
+The categorical product and the usual product agrees
+-/
+def _root_.RingEquiv.piEquivPi (R : ι → Type u) [∀ i, CommRing (R i)] :
+    (∏ (fun i : ι ↦ CommRingCat.of (R i)) : CommRingCat.{u}) ≃+* ((i : ι) → R i) :=
+  let R' : ι → CommRingCat.{u} := fun i : ι ↦ CommRingCat.of (R i)
+  RingEquiv.ofHomInv (piIsoPi R').hom (piIsoPi R').inv (piIsoPi R').3 (piIsoPi R').4
+
+
```
Comment (erdOne) [inline Mathlib/Algebra/Category/Ring/Constructions.lean:262]:
```suggestion

```

**Review by erdOne — COMMENTED** [review]
Can you add this to the module docstring?

**Comment by jjaassoonn** [comment]
> Can you add this to the module docstring?

Done.

File: Mathlib/Algebra/Category/Ring/Constructions.lean [hunk 3]
```diff
@@ -215,6 +216,49 @@ set_option linter.uppercaseLean3 false in
...
+    map_add' := by aesop }
+  fac := by aesop
+  uniq s g h := FunLike.ext _ _ fun x ↦ funext fun i ↦ FunLike.congr_fun (h ⟨i⟩) x
+
+/--
+The categorical product and the usual product agrees
+-/
+def piIsoPi : ∏ R ≅ CommRingCat.of ((i : ι) → R i) :=
+  limit.isoLimitCone ⟨_, piFanIsLimit R⟩
+/--
```
Comment (erdOne) [inline Mathlib/Algebra/Category/Ring/Constructions.lean:253]:
```suggestion

/--
```
oops might be my fault

**Comment by erdOne** [comment]
maintainer merge

File: Mathlib/Algebra/Category/Ring/Constructions.lean [hunk 4]
```diff
@@ -215,6 +216,50 @@ set_option linter.uppercaseLean3 false in
...
+The categorical product of rings is the cartesian product of rings. This is its `Fan`.
+-/
+@[simps! pt]
+def piFan : Fan R :=
+  Fan.mk (CommRingCat.of ((i : ι) → R i)) fun i ↦
+    { toFun := fun g ↦ g i
+      map_one' := by aesop
+      map_mul' := by aesop
+      map_zero' := by aesop
+      map_add' := by aesop }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Category/Ring/Constructions.lean:233]: I think this already exists as `RingHom.eval` or similar?

Comment (erdOne) [reply Mathlib/Algebra/Category/Ring/Constructions.lean:233]: This is `Pi.evalRingHom`.

File: Mathlib/Algebra/Category/Ring/Constructions.lean [hunk 5]
```diff
@@ -215,6 +216,50 @@ set_option linter.uppercaseLean3 false in
...
+
+/--
+The categorical product of rings is the cartesian product of rings.
+-/
+def piFanIsLimit : IsLimit (piFan R) where
+  lift s :=
+  { toFun := fun x i ↦ s.π.app ⟨i⟩ x
```
Comment (erdOne) [inline Mathlib/Algebra/Category/Ring/Constructions.lean:240]: And I think this is `Pi.ringHom`?

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by sgouezel** [comment]
bors r-
Can you merge master and fix the build?
bors d+

**Pull request closed without merging**